Sources/CSRF.swift

```diff
@@ -57,139 +50,59 @@ public struct CSRF: Middleware {
+
+        guard let secret = session["CSRFSecret"] else {
+            let random = CryptoRandom()
+            let secretData = try random.generateData(count: 16)
+            let secret = secretData.hexEncodedString()
+            session["CSRFSecret"] = secret
             return secret
         }
         
         return secret
     }
     
-    private func defaultTokenRetrieval(from request: Request) throws -> String {
-        if let token = request.parameters["_csrf"]?.string {
-            return token
-        }
+    private static func defaultTokenRetrieval(from request: Request) throws -> Future<String> {
         
+        let promise = request.eventLoop.newPromise(String.self)
         let csrfKeys: Set<String> = ["_csrf", "csrf-token", "xsrf-token", "x-csrf-token", "x-xsrf-token", "x-csrftoken"]
-        let requestHeaderKeys = Set(request.headers.keys.map { $0.key })
+        let requestHeaderKeys = Set(request.http.headers.map { $0.name })
         
         let intersection = csrfKeys.intersection(requestHeaderKeys)
         
-        guard let match = intersection.first else {
-            throw Abort(.forbidden,
-                        metadata: nil,
-                        reason: "No CSRF token provided.",
-                        identifier: nil,
-                        possibleCauses: ["Perhaps you forgot to create a token for the session.",
-                                         "You could be using a different header key for the CSRF token than is covered by the default token retrieval."],
-                        suggestedFixes: ["Make sure to create and add a token to your request header. \nSee `CSRF.createToken(from:) for documentation on usage.",
-                                         "Look at `CSRF.defaultTokenRetrieval(from:)` to see what keys are looked for by default."],
-                        documentationLinks: nil,
-                        stackOverflowQuestions: nil,
-                        gitHubIssues: nil)
+        if let matchingKey = intersection.first, let token = request.http.headers[matchingKey].first {
+            promise.succeed(result: token)
+            return promise.futureResult
         }
         
-        let matchingKey = HeaderKey(match)
-        guard let token = request.headers[matchingKey]?.string else {
-            throw Abort(.forbidden,
-                        metadata: nil,
-                        reason: "Failed to find token for key: \(matchingKey).",
-                        identifier: nil,
-                        possibleCauses: nil,
-                        suggestedFixes: nil,
-                        documentationLinks: nil,
-                        stackOverflowQuestions: nil,
-                        gitHubIssues: nil)
+        request.content.get(at: "_csrf").do { token in
+            promise.succeed(result: token)
+        }.catch { _ in
+            promise.fail(error: Abort(.forbidden, reason: "No CSRF token provided."))
```

Comment:
you can use [`catchMap(...)`](https://api.vapor.codes/core/latest/Async/Extensions/Future.html#/s:3NIO15EventLoopFutureC5AsyncE8catchMapACyxGxs5Error_pKcF) to convert the error case.